web-ui/src/streaming/builder/hooks/useAddIoNode.ts

```diff
@@ -0,0 +1,140 @@
+// Logic to add either a input or output node to the graph.
+
+import { useCallback } from 'react'
+import { NodeProps, useReactFlow, getConnectedEdges } from 'reactflow'
+import { AttachedConnector, Direction } from 'src/types/manager'
+import { ConnectorDescr } from 'src/types/manager/models/ConnectorDescr'
+import { Schema } from 'src/types/program'
+import { randomString } from 'src/utils'
+
+const HEIGHT_OFFSET = 120
+
+// Checks if the connector can connect to a give schema
+export function connectorConnects(ac: AttachedConnector, schema: Schema): boolean {
+  console.log('connectorConnects', ac, schema)
+  if (ac.direction === Direction.INPUT) {
+    return schema.inputs.some(table => table.name === ac.config)
+  } else {
+    return schema.outputs.some(view => view.name === ac.config)
+  }
+}
+
+export function useAddConnector() {
+  const { setNodes, getNodes, getNode, addNodes, addEdges } = useReactFlow()
+
+  const addNewConnector = useCallback(
+    (connector: ConnectorDescr, ac: AttachedConnector) => {
+      // Input or Output?
+      const newNodeType = ac.direction === Direction.INPUT ? 'inputNode' : 'outputNode'
+      const placeholderId = ac.direction === Direction.INPUT ? 'inputPlaceholder' : 'outputPlaceholder'
+      const placeholder = getNode(placeholderId)
+      if (!placeholder) {
+        return
+      }
+
+      // If this node already exists, don't add it again
+      const existingNode = getNodes().find(node => node.id === ac.uuid)
+      if (existingNode) {
+        return
+      } else {
+        // Move the placeholder node down a bit
```

Comment:
shouldn't this use flow or something automatic for layout?

Reply:
it will eventually get recomputed, but it looks better if it moves to the correct spot immediately here, added a comment for this.